cmd/kube-proxy/app/server_others.go

```diff
@@ -137,7 +137,8 @@ func newProxyServer(
 	if nodeIP.IsUnspecified() {
 		nodeIP = utilnode.GetNodeIP(client, hostname)
 		if nodeIP == nil {
-			return nil, fmt.Errorf("unable to get node IP for hostname %s", hostname)
+			klog.Warning("invalid nodeIP, initializing kube-proxy with 127.0.0.1 as nodeIP")
+			nodeIP = net.ParseIP("127.0.0.1")
```

Comment:
The proxier implementations already sets node IP to 127.0.0.1 if it doesn't exist already, wondering if it makes more sense to remove this nil check altogether (see https://github.com/kubernetes/kubernetes/pull/84071).

Reply:
This change _may_ now impact `ipvs.NewProxier()` (I can't tell for sure currently as I'm on mobile only) which used to not necessarily come with a set node IP.

Is there a reason we cannot just drop line 140 without adding anything else? That's what the code originally used to do, so it seems like the cleaner approach to me with regards to addressing the bug.

Reply:
IPVS proxier has the same behavior as iptables for nil `nodeIP` (see https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L406-L409) so agreed with @timoreimann this can just be
```go
if nodeIP.IsUnspecified() {
	if nodeIP.IsUnspecified() {
		nodeIP = utilnode.GetNodeIP(client, hostname)
        }
}
```

Reply:
IPVS proxier has the same behavior as iptables for nil `nodeIP` (see https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L406-L409) so agreed with @timoreimann this can just be
```go
if nodeIP.IsUnspecified() {
	if nodeIP.IsUnspecified() {
		nodeIP = utilnode.GetNodeIP(client, hostname)
        }
}
```

Reply:
Should we remove the code in NewProxier implementations that checks for and handles nil?  This is kind of a contract change - they can assume it is non-nil.

So I guess from a compat POV this is correct.  From a correctness POV, this is wrong.  At least the iptables module will make a wrong decision based on this information.  IPVS seems only to log it.

Reply:
Having the user set the --bind-address seems like the RIGHT answer, but I acknowledge that this is a breaking change and we should fix it.

Can we enhance the log message to say something like:
````
    klog.V(0).Infof("can't determine this node's IP, assuming 127.0.0.1; if this is incorrect, set the --bind-address flag")
```

Of course, flags are passe and config files are vogue, and it's harder to reference config file fields.  This is also wrongish in the case of IPv6.

Reply:
> Should we remove the code in NewProxier implementations that checks for and handles nil? This is kind of a contract change - they can assume it is non-nil.

If  iptables.NewProxier was only used here, it might be redundant to validate nodeIP `nil` again. But, I found that it was used by https://github.com/kubernetes/kubernetes/blob/master/pkg/kubemark/hollow_proxy.go#L82, so I withdrew the modification here

Reply:
Can we fix that do explicitly pass 127.0.0.1 and remove it from NewProxier?

Reply:
> Can we fix that do explicitly pass 127.0.0.1 and remove it from NewProxier?

Done.PTAL